[UI] Add metrics tab in Node Page

ChengYanJin opened the pull request:
**Component**: UI

**Context**: 
Add metrics tab in the New Node Page, the default timespan is **Last 24 hours**

**Summary**:
Note that: 
- Miss the selection of timespan, by default is **Last 24 hours**
- Miss the auto-refresh for the graphs

(Will address them it in PR #2809 ).

**Update**:
- Make the Graph Container scrollable in the Metrics tab
- Set the width and height of the graph according to the `innerHeight` and `innerWidth` of the screen in order to leverage the space we have to display the charts

**Acceptance criteria**: 
![image](https://user-images.githubusercontent.com/18453133/93621716-442ad300-f9dc-11ea-99e0-9fb3385a71e3.png)
Update with the line color in graphs with new design:
![image](https://user-images.githubusercontent.com/18453133/94673807-9098db80-0317-11eb-8f93-f50491e6ce74.png)

ChengYanJin committed:
ui/styles: Add the common style for tabs

Refs: #2774

ChengYanJin committed:
ui/ducks/monitoring: Add fetch node stats for the metrics tab

Refs: #2776

ChengYanJin committed:
ui/stles: Adapt the padding because of removing the tab title

Refs: #2776

ChengYanJin committed:
ui/nodepagemetrics: Add six graphs in the metrics tab

Refs: #2776

ChengYanJin committed:
ui/nodepagersp: Fetch the data for metrics tab

Refs: #2776

ChengYanJin committed:
ui/nodepagecontent: Get the current info and pass it to NodePageRSP

Refs: #2776

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -177,6 +208,8 @@ export const isCurrentVolumeStatsRefresh = (state) =>
   state.app.monitoring.volumeCurrentStats.isRefreshing;
 export const metricsTimeSpan = (state) =>
   state.app.monitoring.volumeStats.metricsTimeSpan;
+export const nodeMetricsTimeSpan = (state) =>
```
Since we're adding a `nodeMetricsTimeSpan`, would make sense to rename the one above `volumeMetricsTimeSpan`

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -521,6 +554,174 @@ export function* stopRefreshCurrentStats() {
+    Math.round(currentTime.getTime() / 1000) - sampleDuration;
+  const startingTimeISO = new Date(startingTimestamp * 1000).toISOString();
+
+  const cpuUsageQuery = `(((count(count(node_cpu_seconds_total{instance=~"${instanceIP}:${PORT_NUMBER_PROMETHEUS}"}) by (cpu))) - avg(sum by (mode)(irate(node_cpu_seconds_total{mode='idle',instance=~"${instanceIP}:${PORT_NUMBER_PROMETHEUS}"}[5m])))) * 100) / count(count(node_cpu_seconds_total{instance=~"${instanceIP}:${PORT_NUMBER_PROMETHEUS}"}) by (cpu))`;
```
I was surprised that your queries could work with the Prometheus port, because we get these metrics from `node-exporter`... But indeed, this `PORT_NUMBER_PROMETHEUS` constant is just not using the right name, because `9100` **is** `node-exporter`'s port (Prom is listening on 9090).

gdemonet commented on ui/src/ducks/app/monitoring.js:
```diff
@@ -521,6 +554,174 @@ export function* stopRefreshCurrentStats() {
+    controlPlaneNetworkBandwidthIn: controlPlaneNetworkBandwidthIn,
+    controlPlaneNetworkBandwidthOut: controlPlaneNetworkBandwidthOut,
+    workloadPlaneNetworkBandwidthIn: workloadPlaneNetworkBandwidthIn,
+    workloadPlaneNetworkBandwidthOut: workloadPlaneNetworkBandwidthOut,
```
Did you know the keys are not necessary in ES6 notation when the value is a variable of the same name? It's called ["shorthand property names"](https://alligator.io/js/object-property-shorthand-es6/#:~:text=New%20in%20JavaScript%20with%20ES6,simply%20pass%20the%20key%20name.)

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -1,8 +1,436 @@
+} from '../constants';
+
+const GraphsContainer = styled.div`
+  display: flex;
+  flex-direction: column;
```
Just a suggestion: wouldn't it be easier to use CSS Grid instead of a combination of Flexbox containers? Allows for a simpler gap size management, and uniform sizing (looks better when some of the charts are empty, for instance).

ChengYanJin replied:
Actually I didn't see the big difference when come to use CSS Grid. For me, those two can both solve our issue. Maybe just we are more familiar with Flexbox and use it everywhere in our code.

When some of the charts are missing, we'll have an empty chart today. But we should display an error message which is missing today.

gdemonet replied:
Won't fight on this topic, just suggesting you try Grid a bit more for building "grid layouts" (which is the case here). There's a reason they built it ;)

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -1,8 +1,436 @@
+      ticks: true,
+      tickCount: 4,
+      labelAngle: -50,
+      labelColor: '#a8b5c1',
```
Shouldn't we have all colors in the theme? (same for chart lines)

ChengYanJin replied:
Indeed, the color of the label should come from the `defaultTheme`. 
However, the color of the chart lines are exceptional color and we don't use them anywhere else yet. For me, it makes sense to hardcode the color today.

gdemonet replied:
I'm not sure chart lines are "exceptional" in our UIs, since most of our resources will have some form of monitoring metrics. We can hardcode them today, sure, but it'd be a good idea to open a ticket in core-ui and think about it more.

ChengYanJin replied:
Sure! I will open a ticket in core-ui. 
And actually, the colors we use in the graphs is still an open topic we haven't finalized yet. We still need to work with PO & UX.

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -87,98 +85,98 @@ const NodePageMetricsTab = (props) => {
-  });
-  const nodeControlPlaneNetworkBandwidthInData = controlPlaneNetworkBandwidthInOperated?.map(
-    (slot) => {
+  const cpuUsageData = nodeStatsOperated
+    ?.find((obj) => obj.metrics === 'cpuUsage')
+    ?.data?.map((slot) => {
+      return {
+        date: fromUnixTimestampToDate(slot[0]),
+        y: slot[1],
+      };
+    });
```
Weren't trying to reduce duplication? :stuck_out_tongue_winking_eye:

Instead of doing this mapping after building `nodeStatsOperated`, you could do it during construction:

```javascript
  const nodeStatsOperated = Object.keys(nodeStats).reduce((acc, metricName) => {
    const data = operateMetricRawData(nodeStats[metricName][0]?.values);
    acc[metricName] = data.map((slot) => ({
      date: fromUnixTimestampToDate(slot[0]),
      y: slot[1],
    }));
    return acc;
  }, {});
```

And then consume it like:

```javascript
<LineChart
  id={'node_cpu_usage_id'}
  data={nodeStatsOperated['cpuUsage']}
  {...rest}
/>
```

ChengYanJin replied:
Hello @gdemonet 
Was trying to reduce the duplication here too. But we have different return keys when it comes to the graphs with multiple line (in/out, read/write)
Please see:
```
  const nodeWorkloadPlaneNetworkBandwidthOutData = nodeStatsOperated
    ?.find((obj) => obj.metrics === 'workloadPlaneNetworkBandwidthOut')
    ?.data?.map((slot) => {
      return {
        date: fromUnixTimestampToDate(slot[0]),
        out: slot[1],
        type: 'out',
      };
    });
```

gdemonet replied:
Oh right, missed that one. There still are some ways to reduce this:

- each `.find((obj) => obj.metrics === someKey)` could be simplified if `nodeStatsOperated` was a map instead of an array:

```javascript
const nodeWorkloadPlaneNetworkBandwidthOutData = nodeStatsOperated
  ?.['workloadPlaneNetworkBandwidthOut']?.map((slot) => { ... })
```

- if you want to push it further, adding a `type` key can be done easily by inferring it from the `metricName`:

```javascript
// First variant, assumes a given format for `metricName`
const nodeStatsOperated = Object.keys(nodeStats).reduce((acc, metricName) => {
  const data = operateMetricRawData(nodeStats[metricName][0]?.values);

  let extra = {};
  if (metricName.endsWith('Out')) extra.type = 'out';
  else if (metricName.endsWith('In')) extra.type = 'in';
  else if (metricName.endsWith('Read')) extra.type = 'read';
  else if (metricName.endsWith('Write')) extra.type = 'write';

  acc[metricName] = data.map((slot) => ({
    date: fromUnixTimestampToDate(slot[0]),
    y: slot[1],
    ...extra
  }));
  return acc;
}, {});

// Second variant, stores the mapping between types and metric names
const typedMetrics = {
  iopsRead: 'read',
  iopsWrite: 'write',
  controlPlaneNetworkBandwidthIn: 'in',
  controlPlaneNetworkBandwidthOut: 'out',
  workloadPlaneNetworkBandwidthIn: 'in',
  workloadPlaneNetworkBandwidthOut: 'out',
};

const nodeStatsOperated = Object.keys(nodeStats).reduce((acc, metricName) => {
  const data = operateMetricRawData(nodeStats[metricName][0]?.values);

  let extra = {};
  const metricType = typedMetrics[metricName];
  if (metricType !== undefined) extra.type = metricType;

  acc[metricName] = data.map((slot) => ({
    date: fromUnixTimestampToDate(slot[0]),
    y: slot[1],
    ...extra
  }));
  return acc;
}, {});
```

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -406,7 +404,7 @@ const NodePageMetricsTab = (props) => {
               />
             ) : (
-              <NoDataGraphText>Loading…</NoDataGraphText>
+              <LoaderContainer size="small"></LoaderContainer>
```
Using a loader instead of some text isn't related to the "code deduplication" commit, would be better to create another one.

gdemonet reviewed (changes requested):
Some code review comments :wink: 

Also, I played with the page, and have a few comments:

- If my screen is too short in height, I can't see the bottom graphs **and I can't scroll**
- If my screen is too short in width, the graphs are going over the table (I'd rather have scrollbar and push them to the right)
- Whichever the size of my display or the level of zoom, I have (ugly, but that's Chrome's fault) scrollbars everywhere - this problem is true on all pages actually, am I the only one @scality/metalk8s?:

[![2Agodv.md.png](https://iili.io/2Agodv.md.png)](https://freeimage.host/i/2Agodv)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @tcarmet

* @nootal

* @teddyandrieux

* @lucieleonard

* @gdemonet

* @sayf-eddine-scality

* @nicolast

* @jbertran

* @monpote

* @carlito-scality

* @slaperche-scality

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

ChengYanJin committed:
ui/nodes: Rename the constant PORT_NUMBER_PROMETHEUS to PORT_NODE_EXPORTER

Refs: #2776

ChengYanJin committed:
ui/ducks/monitoring: Rename the volumeMetricsTimeSpan selector

Since we have metricsTimeSpan for both of volume and node, it's better to rename the two selectors

Refs: #2776

ChengYanJin committed:
ui/ducks/monitoring: Remove the key when the value is a variable of the same name

According to ES6 Notation.

Refs: #2776

ChengYanJin committed:
ui: Use the textSecondary color for the x-axis label in the graphs

Refs: #2776

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -87,98 +85,98 @@ const NodePageMetricsTab = (props) => {
+    });
+
+  const nodeIOPSWriteData = nodeStatsOperated
+    ?.find((obj) => obj.metrics === 'iopsRead')
```
Looks like you switched them up, no?

```suggestion
    ?.find((obj) => obj.metrics === 'iopsWrite')
```

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -87,98 +85,98 @@ const NodePageMetricsTab = (props) => {
+    });
+
+  const nodeIOPSReadData = nodeStatsOperated
+    ?.find((obj) => obj.metrics === 'iopsWrite')
```
```suggestion
    ?.find((obj) => obj.metrics === 'iopsRead')
```

ChengYanJin committed:
ui/package: Update the core-ui version

Add a new prop activeColor for Tabs component

Refs: #2776

ChengYanJin committed:
ui/nodes: Reduce the duplication when modify the data for the linechart

Refs: #2776

ChengYanJin committed:
ui/nodes: Extract the common y-axis specification for vega-lite linechart

Refs: #2776

ChengYanJin committed:
ui/nodes: Reduce the duplication when operate the raw data for the graph

Refs: #2776

ChengYanJin committed:
ui/nodes: Replace the loading text with the <Loader> component

ChengYanJin committed:
ui/volumes: Use the extracted the linechart spec for y-axis

Refs: #2776

ChengYanJin committed:
ui/styles: Add overflow hidden to the PageContainer to avoid the ugly scroll

Refs: #2776

ChengYanJin committed:
ui/nodes: Make the Graph container inside Metrics tab scrollable

Refs: #2776

ChengYanJin committed:
ui/nodes: Set the height and width of the graph according to the current size of the screen

ChengYanJin committed:
ui/ducks/monitoring: Initialize node metrics with array instead of object

Refs: #2776

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/2776-add-metrics-graph-in-node-page` and the
destination branch `development/2.6`.

Please resolve the conflict on **the feature branch** (`improvement/2776-add-metrics-graph-in-node-page`).

```bash
 $ git fetch
 $ git checkout origin/improvement/2776-add-metrics-graph-in-node-page
 $ git merge origin/development/2.6
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/2776-add-metrics-graph-in-node-page
```

gdemonet commented on ui/src/components/LinechartSpec.js:
```diff
@@ -0,0 +1,40 @@
+export const yAxisUsauge = [
```
typo:

```suggestion
export const yAxisUsage = [
```

gdemonet commented on ui/src/components/NodePageMetricsTab.js:
```diff
@@ -1,8 +1,325 @@
+import { LineChart, Loader } from '@scality/core-ui';
 import { TabContainer } from './CommonLayoutStyle';
+import {
+  yAxisUsauge,
```
```suggestion
  yAxisUsage,
```

gdemonet commented on ui/src/components/VolumeMetricGraphCard.js:
```diff
@@ -16,6 +16,7 @@ import {
   addMissingDataPoint,
   fromUnixTimestampToDate,
 } from '../services/utils';
+import { yAxisUsauge, yAxisWriteRead } from './LinechartSpec';
```
```suggestion
import { yAxisUsage, yAxisWriteRead } from './LinechartSpec';
```

gdemonet reviewed (dismissed):
Just a small typo, should be good otherwise :)

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2776-add-metrics-graph-in-node-page` into
`development/2.6`
* [w/2.7/improvement/2776-add-metrics-graph-in-node-page](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/2776-add-metrics-graph-in-node-page) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

ChengYanJin committed:
ui/nodes: Update the graph line colors with the new design

Refs: #2776

gdemonet reviewed (approved):
:shipit:

ChengYanJin commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/010fada421e73e0ad030a9bf8baf782a347dc3e6) did not succeed in branch w/2.7/improvement/2776-add-metrics-graph-in-node-page.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/18404ad5617311d4aaa742665289e4b3a2ee873a) did not succeed in branch improvement/2776-add-metrics-graph-in-node-page.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/010fada421e73e0ad030a9bf8baf782a347dc3e6) did not succeed in branch w/2.7/improvement/2776-add-metrics-graph-in-node-page.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye chengyanjin.
